This is an exploratory analysis of heart disease patients obtained from Codecademy (Data Science pathway) which in turn was obtained from https://archive.ics.uci.edu/ml/datasets/heart+disease

Begin by importing the data into a data frame.

In [44]:
import pandas as pd
import numpy as np

patients_df = pd.read_csv('processed.cleveland.data.csv')
patients_df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  heart_disease  
0    3.0  0.0  6.0              0  
1    2.0  3.0  3.0              2  
2    2.0  2.0  7.0              1  
3    3.0  0.0  3.0              0  
4    1.0  0.0  3.0              0

Data dictionary descriptions for the fields:

    age: age in years
    sex: sex (1 = male; 0 = female)
    cp: chest pain type
        -- Value 1: typical angina
        -- Value 2: atypical angina
        -- Value 3: non-anginal pain
        -- Value 4: asymptomatic
    trestbps: resting blood pressure (in mm Hg on admission to the hospital)
    chol: serum cholestoral in mg/dl
    fbs: (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
    restecg: resting electrocardiographic results
        -- Value 0: normal
        -- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
        -- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
    thalach: maximum heart rate achieved
    exang: exercise induced angina (1 = yes; 0 = no)
    oldpeak = ST depression induced by exercise relative to rest
    slope: the slope of the peak exercise ST segment
        -- Value 1: upsloping
        -- Value 2: flat
        -- Value 3: downsloping
    ca: number of major vessels (0-3) colored by flourosopy
    thal: 3 = normal; 6 = fixed defect; 7 = reversable defect
    num (renamed heart_disease): diagnosis of heart disease (angiographic disease status)
        -- Value 0: < 50% diameter narrowing
        -- Value 1: > 50% diameter narrowing
        (in any major vessel: attributes 59 through 68 are vessels)

In [45]:
patients_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   age            303 non-null    float64
 1   sex            303 non-null    float64
 2   cp             303 non-null    float64
 3   trestbps       303 non-null    float64
 4   chol           303 non-null    float64
 5   fbs            303 non-null    float64
 6   restecg        303 non-null    float64
 7   thalach        303 non-null    float64
 8   exang          303 non-null    float64
 9   oldpeak        303 non-null    float64
 10  slope          303 non-null    float64
 11  ca             303 non-null    object 
 12  thal           303 non-null    object 
 13  heart_disease  303 non-null    int64  
dtypes: float64(11), int64(1), object(2)
memory usage: 33.3+ KB


ca and thal are object types, most likely strings, but the dictionary list them as int values indicating that there may be some non-numeric data points.

In [46]:
print(patients_df.ca.unique())
print(patients_df.thal.unique())

['0.0' '3.0' '2.0' '1.0' '?']
['6.0' '3.0' '7.0' '?']


Each of the columns contains a '?' value, probably for unknown values.  

Change to NaN values to convert column to numeric.

In [47]:
patients_df = patients_df.replace('?', np.nan)
print(patients_df.ca.unique())
print(patients_df.thal.unique())

['0.0' '3.0' '2.0' '1.0' nan]
['6.0' '3.0' '7.0' nan]


In [48]:
patients_df[patients_df.ca=='?']

Empty DataFrame
Columns: [age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, heart_disease]
Index: []

In [49]:
patients_df[patients_df.thal=='?']

Empty DataFrame
Columns: [age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, thal, heart_disease]
Index: []

In [50]:
patients_df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  heart_disease  
0    3.0  0.0  6.0              0  
1    2.0  3.0  3.0              2  
2    2.0  2.0  7.0              1  
3    3.0  0.0  3.0              0  
4    1.0  0.0  3.0              0

Convert sex to male/female
sex: sex (1 = male; 0 = female)

In [52]:
patients_df.sex.replace({1:"male", 0:"female"}, inplace=True)
patients_df.head()

age     sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0    male  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0    male  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0    male  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0    male  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  female  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  heart_disease  
0    3.0  0.0  6.0              0  
1    2.0  3.0  3.0              2  
2    2.0  2.0  7.0              1  
3    3.0  0.0  3.0              0  
4    1.0  0.0  3.0              0

Rename and convert cp values
cp: chest pain type
    -- Value 1: typical angina
    -- Value 2: atypical angina
    -- Value 3: non-anginal pain
    -- Value 4: asymptomatic

In [59]:
patients_df.rename(columns={'cp':'chest_pain'}, inplace=True)
patients_df.chest_pain.replace({1:'typical angina', 2:'atypical angina', 3:'non-anginal pain', 4:'asymptomatic'}, inplace=True)
patients_df.head()

age     sex        chest_pain  trestbps   chol  fbs  restecg  thalach  \
0  63.0    male    typical angina     145.0  233.0  1.0      2.0    150.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0      2.0    108.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0      2.0    129.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0      0.0    187.0   
4  41.0  female   atypical angina     130.0  204.0  0.0      2.0    172.0   

   exang  oldpeak  slope   ca thal  heart_disease  
0    0.0      2.3    3.0  0.0  6.0              0  
1    1.0      1.5    2.0  3.0  3.0              2  
2    1.0      2.6    2.0  2.0  7.0              1  
3    0.0      3.5    3.0  0.0  3.0              0  
4    0.0      1.4    1.0  0.0  3.0              0

convert restecg
restecg: resting electrocardiographic results
    -- Value 0: normal
    -- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
    -- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria

In [60]:
patients_df.restecg.replace({0:'normal', 1:'ST-T wave abnormality', 2:'left ventricular hypertrophy'}, inplace = True)
patients_df.head()

age     sex        chest_pain  trestbps   chol  fbs  \
0  63.0    male    typical angina     145.0  233.0  1.0   
1  67.0    male      asymptomatic     160.0  286.0  0.0   
2  67.0    male      asymptomatic     120.0  229.0  0.0   
3  37.0    male  non-anginal pain     130.0  250.0  0.0   
4  41.0  female   atypical angina     130.0  204.0  0.0   

                        restecg  thalach  exang  oldpeak  slope   ca thal  \
0  left ventricular hypertrophy    150.0    0.0      2.3    3.0  0.0  6.0   
1  left ventricular hypertrophy    108.0    1.0      1.5    2.0  3.0  3.0   
2  left ventricular hypertrophy    129.0    1.0      2.6    2.0  2.0  7.0   
3                        normal    187.0    0.0      3.5    3.0  0.0  3.0   
4  left ventricular hypertrophy    172.0    0.0      1.4    1.0  0.0  3.0   

   heart_disease  
0              0  
1              2  
2              1  
3              0  
4              0